modules/layouts/LayoutCS.chpl

```diff
@@ -0,0 +1,656 @@
+    }
+
+    if nnz == 0 {
+
+      const dupCount = if isUnique then 0 else _countDuplicates(inds);
+
+      nnz += inds.size-dupCount;
+      _bulkGrow();
+
+      var idxIdx = 1;
+      var prevIdx = parentDom.low - (1,1);
+      var current: idxType;
+
+      if this.row then
+        current = parentDom.dim(1).low;
+      else
+        current = parentDom.dim(2).low;
+
+      // Update startIdx && idx
+      for (i,j) in inds {
+        if !isUnique && (i,j) == prevIdx then continue;
+        else prevIdx = (i,j);
+
+        if this.row {
+          while i != current {
+            current += 1;
+            startIdx[current+1] = startIdx[current];
+          }
+          startIdx[i+1] += 1;
+          idx[idxIdx] = j;
+          idxIdx += 1;
+        } else {
+          while j != current {
+            current += 1;
+            startIdx[current+1] = startIdx[current];
+          }
+          startIdx[j+1] += 1;
+          idx[idxIdx] = i;
+          idxIdx += 1;
+        }
+      }
+
+
+      // make sure startIdx[i]>startIdx[j] for i>j for possibly
+      // untouched part of startIdx
+      const startIdxHigh = startIdx[current+1];
+      for r in current+2..startIdx.domain.high {
+        startIdx[r] = startIdxHigh;
+      }
+
+      return idxIdx-1;
+    } // if nnz == 0
+
+    const (actualInsertPts, actualAddCnt) =
+      __getActualInsertPts(this, inds, isUnique);
+
+    const oldnnz = nnz;
+    nnz += actualAddCnt;
+
+    // Grow nnzDom if necessary
+    _bulkGrow();
+
+    // Linearly fill the new idx from backwards
+    var newIndIdx = indsDom.high; //index into new indices
+    var oldIndIdx = oldnnz; //index into old indices
+    var newLoc = actualInsertPts[newIndIdx]; // its position-to-be in new dom
+    while newLoc == -1 {
+      newIndIdx -= 1;
+      if newIndIdx == indsDom.low-1 then break; // there were duplicates -- now done
+      newLoc = actualInsertPts[newIndIdx];
+    }
+
+    var arrShiftMap: [{1..oldnnz}] int; //to map where data goes
+
+    for i in 1..nnz by -1 {
+      if oldIndIdx >= 1 && i > newLoc {
+        // Shift from old values
+        idx[i] = idx[oldIndIdx];
+        arrShiftMap[oldIndIdx] = i;
+        oldIndIdx -= 1;
+      }
+      else if newIndIdx >= indsDom.low && i == newLoc {
+        // Put the new guy in
+        if this.row {
+          idx[i] = inds[newIndIdx][2];
+        } else {
+          idx[i] = inds[newIndIdx][1];
+        }
+        newIndIdx -= 1;
+        if newIndIdx >= indsDom.low then
+          newLoc = actualInsertPts[newIndIdx];
+        else
+          newLoc = -2; // Finished new set
+        while newLoc == -1 {
+          newIndIdx -= 1;
+          if newIndIdx == indsDom.low-1 then break; // There were duplicates -- now done
+          newLoc = actualInsertPts[newIndIdx];
+        }
+      }
+      else halt("Something went wrong");
+    }
+
+    // Aggregated row || col shift
+    var prevCursor = if this.row then parentDom.dim(1).low else parentDom.dim(2).low;
```

Comment:
It seems as though, if we set up constants like 'compressedDim' and 'indexedDim' (to pick some potentially lame names) that store 1 and 2 (or 2 and 1) depending on whether we were doing CSR or CSC, a bunch of code like this could be simplified from using conditionals to simpler expressions like 'var prevCursor = parentDom.dim(compressedDim)'.